<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good)</span></div>

## 1 Interrupting the flow

In [1]:
for i in range(25):
    if i == 13:
        continue
    if i == 17:
        print("I don't like 17...")
        break
    print(i)
print("And we're done!")

0
1
2
3
4
5
6
7
8
9
10
11
12
14
15
16
I don't like 17...
And we're done!


## 2 List comprehension!

### 2.1 Basic syntax

In [2]:
first_list = [i for i in range(1, 31, 2)]
second_list = [10*i for i in first_list]
print(second_list)

[10, 30, 50, 70, 90, 110, 130, 150, 170, 190, 210, 230, 250, 270, 290]


### 2.2 List comprehension with conditions

In [3]:
first_list = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32]
second_list = [i**2 for i in first_list if (i % 2 == 0 and i <28)]

print(second_list)

[4, 16, 36, 64, 100, 144, 196, 256, 324, 400, 484, 576, 676]


## 3 Other useful stuff

### 3.1 for with unpacking

In [4]:
x, y, z = [1, 2, 3]
print(f'x = {x}, y = {y}, z = {z}')
#this allows you to make a list as a dumpground for values with names.
#used only when you don't care about the main list or its length and you never need to refer to the whole list
#since the actual list itself is impossible to call (since it's unnamed)
#the benefit this offers is that you can easily have names for items without having to make a ton of different variables

'''
x, y, z = 1, 2, 3
print(f'x = {x}, y = {y}, z = {z}')         #this is another way to do the same thing, but you'd have to make many variables
'''
print()        #i don't want jupyter to output the comment above

x = 1, y = 2, z = 3



### 3.2 for with a zip()

In [5]:
super_names = ["Black Widow", "Iron Man", "Doctor Strange"]
real_names = ["Natasha Romanoff", "Tony Stark", "Stephen Strange"]
movie_release_dates =[2021, 2008, 2016]

for real_name, super_name, release_date in zip(real_names,super_names, movie_release_dates):
    print(f"{real_name} is Marvel's {super_name}! Their premeire movie came out in {release_date}.")
    
a = zip(real_names, super_names, movie_release_dates)
print(type(a))

Natasha Romanoff is Marvel's Black Widow! Their premeire movie came out in 2021.
Tony Stark is Marvel's Iron Man! Their premeire movie came out in 2008.
Stephen Strange is Marvel's Doctor Strange! Their premeire movie came out in 2016.
<class 'zip'>


### 3.3 for with dictionaries

In [6]:
super_names = ["Black Widow", "Iron Man", "Doctor Strange"]
real_names = ["Natasha Romanoff", "Tony Stark", "Stephen Strange"]
super_aliases = {}
for super_name, real_name in zip(super_names, real_names):
    super_aliases[super_name]= real_name

for super_name, real_name in super_aliases.items():
    print(f"{super_name}'s real identity is {real_name}")

Black Widow's real identity is Natasha Romanoff
Iron Man's real identity is Tony Stark
Doctor Strange's real identity is Stephen Strange


## Exercise 1 :  Make me an odd list

In [7]:
%%timeit
odd_list = []
for i in range(1,10,2):
    odd_list.append(i**2)

1.34 µs ± 17 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## Exercise 2 :  Make me another odd list

In [8]:
%%timeit
odd_list = [i**2 for i in range(1,10,2)]

1.3 µs ± 15.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## Exercise 3 :  Time me!

Seems like list comprehension is very slightly faster, at 1.3 microseconds per loop compared to the 1.34 microseconds of the loop.